Sources/BraintreePayPal/BTPayPalRequest.swift

```diff
@@ -157,6 +136,75 @@ import BraintreeCore
 
     // MARK: Internal Methods
 
+    func parameters(with configuration: BTConfiguration) -> [String: Any] {
+        let baseParameters: [String: Any] = baseParameters(with: configuration)
+
+        switch paymentType {
+        case .checkout:
+            // We downcast self to avoid an empty implementation in the parent class
+            // which would require us to override this function
```

Comment:
We _are_ being a little naughty here. We were trying to avoid writing an "empty" implementation in the super class. We wanted to avoid writing this function:

```swift
func parameters(with configuration: BTConfiguration) -> [String: Any] {
    return [:]
}
```
and overriding it in both classes, since there is no way to _not_ implement it. Though, that solution is probably more standard.

Reply:
I think to also add to this, from my perspective the above is less of a concern than having the logic all in 1 place to build the parameters dictionary. With the above (returning an empty dictionary and overriding it) we have the logic of how we construct the `parameters` living in `BTPayPalRequest` (the base dictionary), as well as `BTPayPalCheckoutRequest` and `BTPayPalVaultRequest` for appending the request specific pieces to the base parameters. This makes it harder to reason about what is being done where to construct the `parameters`.

Having the entirety of how we construct `parameters` live in one place makes it much easier to reason about how we are constructing this dictionary, imo. But of folks feel strongly about returning an empty dictionary here and overriding it in the other files we can do that instead!

Reply:
Hm, I think this violates "separation/responsibility" of concern, since BTPayPalRequest is doing more than it should be.

We could do something like [this](https://github.com/braintree/braintree_ios/compare/convert-paypal-client-accountnonce...refactor-paramaters-override). The superclass implementation is not empty, and each subclass just appends what it needs to on-top of the parent class implementation.

Reply:
Good point - yeah I think I was thinking of it in a weird way initially, but the diff you sent makes sense. I'll make this update and push up the changes. Thanks for talking through that 🙏 !

Reply:
This is the best solution for sure!